### PR TITLE
PMM 3399 QAN API

### PR DESCRIPTION
**Pull request by askomorokhov**
https://jira.percona.com/browse/PMM-3399

### CODE REVIEW

File: api/Gopkg.lock [hunk 1]
```diff
@@ -214,7 +214,6 @@
   ]
   pruneopts = "T"
   revision = "aa810b61a9c79d51363740d207bb46cf8e620ed5"
-  version = "v1.2.0"
```
Comment (AlekSi) [inline api/Gopkg.lock]: Why?

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

File: api/Gopkg.lock (hunk 1 above)
Comment (askomorokhov) [reply api/Gopkg.lock]: Reverted.

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

File: api/qan/version.proto [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+option go_package = "qan";
+
+import "google/api/annotations.proto";
+
+// Version service to check version of QAN-API.
+service Version {
+  // HandleVersion return build version of QAN-API.
+  rpc HandleVersion(VersionRequest) returns (VersionReply) {
+    option (google.api.http) = {
+      post: "/v1/qan/GetReport"
```
Comment (AlekSi) [inline api/qan/version.proto]: Duplicate path.

Let's drop the whole service.

Comment (askomorokhov) [reply api/qan/version.proto]: I need it temporary just to initial test, remove it soon.

Comment (AlekSi) [reply api/qan/version.proto]: Then just fix path. Right now that API can't be used:
```
2019/02/27 16:12:08 args[1:] = [qan/collector.swagger.json qan/metrics.swagger.json qan/profile.swagger.json qan/version.swagger.json]
2019/02/27 16:12:08 paths entry '/v1/qan/GetReport' already exists in primary or higher priority mixin, skipping
```

Comment (askomorokhov) [reply api/qan/version.proto]: Fixed.

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

File: api/qan/collector.proto [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+  string agent_uuid = 9; // Identifier of agent that collect and send metrics.
+  int64 period_start = 10; // Time when collection of bucket started.
+  uint32 period_length = 11; // Duration of bucket.
+  string metrics_source = 12; // Ex.: slowlog, perf schema, Mongo DB, Percona Server, etc.
+  string example = 13; // One of query example from set found in bucket.
+  // ExampleFormat is formant of query example: real or query without values.
+  enum ExampleFormat {
+    QUERY_CLASS_EXAMPLE_FORMAT_INVALID = 0;
+    EXAMPLE = 1;
+    DIGEST = 2;
```
Comment (AlekSi) [inline api/qan/collector.proto]: Should it be FINGERPRINT now? Other fields were renamed to fingerprint.

File: api/qan/metrics.proto [hunk 4]
```diff
@@ -6,54 +6,58 @@ option go_package = "qan";
...
+  repeated string d_client_hosts = 7;
+  string first_seen = 8;
+  string last_seen = 9;
+  map<string, string> labels = 10;
+  string agent_uuid = 11;
+  int64 period_start = 12;
+  uint32 period_length = 13;
+  string example = 14;
   // ExampleFormat is formant of query example: real or query without values
   enum ExampleFormat {
```
Comment (AlekSi) [inline api/qan/metrics.proto]: Duplicate of https://github.com/percona/pmm/pull/268/files#diff-283a55e2672ed0140ef30d059c3be9e0R37.

File: api/qan/collector.proto [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  }
+  ExampleFormat example_format = 14;
+  uint32 is_truncated = 15; // Indicates if query examples is too long and was truncated.
+  // ExampleType is a type of query example selected for this query class in given period of time.
+  enum ExampleType {
```
Comment (AlekSi) [inline api/qan/collector.proto]: Duplicate of https://github.com/percona/pmm/pull/268/files#diff-283a55e2672ed0140ef30d059c3be9e0R45.

File: api/qan/collector.proto [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+// AgentMessage contains several Query Classes for one or more periods of time.
+message AgentMessage {
+  repeated QueryClass query_class = 1;
+}
+
+// QueryClass is aggregated message created by qan-agent.
+// Contains information about one query selected in defined way from query class in specific period of time.
+message QueryClass {
```
Comment (rnovikovP) [inline api/qan/collector.proto]: for now we are not focused on Queries only. Let's have this  as  "Data bucket " or something like this. but this is not about  Olny Queries now

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by AlekSi** [comment]
Is it ready for review?

File: api/qan/collector.proto (hunk 3 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: Renamed.

File: api/qan/metrics.proto (hunk 4 above)
Comment (askomorokhov) [reply api/qan/metrics.proto]: Moved to qan.proto

File: api/qan/collector.proto (hunk 5 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: Moved to qan.proto

File: api/qan/collector.proto (hunk 6 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: Agree. Renamed.

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-268/

**Comment by AlekSi** [comment]
Will resolve conflicts.

**Pull request closed without merging**